text/0095-nft-standard.md

```diff
@@ -0,0 +1,403 @@
+        let symbol: String = storage::read(symbol_uref)
+            .unwrap_or_revert()       // read() -> Result<Option<T>, _>
+            .unwrap_or_revert();      // Option<T> -> T (revert if None)
+    
+        runtime::ret(CLValue::from_t(symbol).unwrap_or_revert());
+    }
+    ```
+
+## Token Metadata Extension
+
+[token-metadata-extension]: #token-metadata-extension
+
+### Hybrid Approach for Token Metadata in CEP-95
+
+[hybrid-approach-for-token-metadata-in-cep-95]: #hybrid-approach-for-token-metadata-in-cep-95
+
+The CEP-95 standard provides a flexible, hybrid approach for NFT token metadata, allowing developers and consumers to leverage **both on-chain and off-chain metadata sources**. Metadata is stored as key-value pairs (`Vec<(String, String)>`), giving a clear yet flexible structure without enforcing a specific serialization format.
+
+- **On-chain metadata**:
+  - Contract stores all its NFT's metadata onchain.
+  - Typically, it its dictionary by `token_metadata` Uref on a contract level where key is a `token_id` and value is a metadata itself.
+- **Off-chain metadata**:
+  - Additionally, CEP-95 supports traditional off-chain metadata retrieval via a URI.
+  - Contracts can provide a `token_uri` within its onchain metadata that itself links to external metadata (e.g., hosted on IPFS or HTTP servers).
+  - This off-chain approach reduces on-chain storage costs and allows easier updates of non-critical metadata fields.
+
+When metadata is created or updated, the contract emits a `MetadataUpdate` event, ensuring that indexers and clients are immediately informed and can cache or display this data without additional lookups.
+
+Consumers, such as wallets, explorers, and marketplaces, benefit from this hybrid approach by first attempting to retrieve on-chain metadata for immediate trustworthiness and availability. If on-chain metadata is unavailable or limited, clients can fall back on the provided metadata URI. This method offers the best balance between efficiency, flexibility, trust, and ease of use, ensuring a smooth and user-friendly NFT experience within the Casper ecosystem.
+
+### Token Metadata Interface: CEP95MetadataEvent
+
+[token-metadata-interface-cep95metadataevent]: #token-metadata-interface-cep95metadataevent
+
+```rust
+pub enum CEP95MetadataEvent {
+    /// Emitted whenever on-chain metadata for a token is created or updated.
+    ///
+    /// @param token_id - The ID of the NFT whose metadata was updated
+    /// @param metadata - The raw metadata (e.g., JSON bytes)
+    MetadataUpdate {
+        token_id: U256,
+    },
+}
+
+/// Optional interface for retrieving token metadata
+pub trait CEP95TokenMetadata {
+    /// Returns metadata for a given token ID.
+    ///
+    /// @param token_id - The NFT ID.
+    /// @return Vec<(String, String)> - metadata stored as a list of key-value pairs.
+    fn token_metadata(&self, token_id: U256) -> Vec<(String, String)>;
```

Comment:
@zie1ony @davidatwhiletrue I propose to move the method above in the main standard interface. If we worry that thaere will be use-cases for tokens w/o metadata, we could wrap the return type in `Option`. My motivation is the following:
1. It enforces the metadata type (a map), which is better for indexing and displaying the assets in network explorers and wallets.
2. It collocates the metadata related event together with the rest of the events, thus makes it easier to see it. I suspect that many readers will focus mostly on the main standard interface.

@Volodymyr-Kuchinskyi please don't force push, because it's killing my comments

Reply:
I think it's a good idea. I'd also use `Option<>` in the return type.

Reply:
`Option` doesn't really make sense here. It's better to return empty vector, even for token id that doesn't exist.

Reply:
@Volodymyr-Kuchinskyi it seems that the consensus  here is to merge the NFT metadata interface into the main contract interface. Let's do that